File: AlgoExpert/Node Depths/solution3.go
package main

type BinaryTree struct {
	Value       int
	Left, Right *BinaryTree
}

type Node struct {
	Node  *BinaryTree
	Depth int
}

func NodeDepths(root *BinaryTree) (sum int) {
	stack := []Node{{Node: root, Depth: 0}}
	var current Node

	for len(stack) > 0 {
		current, stack = stack[len(stack)-1], stack[:len(stack)-1]
		if current.Node != nil {
			sum += current.Depth
			stack = append(stack, Node{Node: current.Node.Left, Depth: current.Depth + 1})
			stack = append(stack, Node{Node: current.Node.Right, Depth: current.Depth + 1})
		}
	}
	return
}

File: AlgoExpert/Node Depths/solution3.py
def nodeDepths(root):
    stack = [(root, 0)]
    depth = 0
    depths_sum = 0
    while stack:
        node, depth = stack.pop()
        if node:
            depths_sum += depth
            stack.extend([(node.left, depth + 1), (node.right, depth + 1)])
    return depths_sum


# This is the class of the input binary tree.
class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
